File: coroutine.pl
#!/usr/bin/perl

use warnings;
use strict;

use Time::HiRes qw(time);
use List::Util  qw(sum);
use Coro;

sub workfunc {
    my ($id, $init) = @_;

    printf "Coroutine %d :: started at %s\n", $id, scalar localtime;
    my $start = time;
    for ( 1 .. 10_000_000 ) {
        $init += ( $_ % 4 ) * ( $_ % 4 ) / ( $_ % 10 + 1 );
        cede;
    }
    printf "Coroutine %d :: stopped at %s\n", $id, scalar localtime;

    terminate time - $start;
}

my $coro_count = shift || 2;
print "Running with $coro_count coroutines\n";

my $master_start = time;
my @coros =
  map { async \&workfunc, $_, rand 10 } 1 .. $coro_count;

my $child_time  = sum map { $_->join() } @coros;
my $master_time = time - $master_start;

printf "Master logged %.3f runtime\n", $master_time;
printf "Coroutines reported %.3f combined runtime\n", $child_time;

File: coroutine.py
from threading import Thread
from random    import random
from time      import asctime, time
from sys       import argv

child_times = []

def worker ( id, init ):
    start = time()
    print "Coroutine %d :: started at %s" % ( id, asctime() )

    for i in xrange( 1, 10000001 ):
        init += ( i % 4 ) * ( i % 4 ) / ( i % 10 + 1 )

    child_times.append( time() - start )
    print "Coroutine %s :: stopped at %s" % ( id, asctime() )

child_count = 2
try:    child_count = int( argv[1] )
except: pass

print "Running with %d coroutines" % (child_count)
child_count += 1 # increment for range

children = []
start    = time()

children = [ Thread( target = worker,
                     args   = ( i, random() * 10 )
                     ) for i in range( 1, child_count ) ]

for c in children: c.start()
for c in children: c.join()

print "Master logged %.3f runtime" % (time() - start)
print "Coroutines reported %.3f combined runtime" % (sum(child_times))
